### PR TITLE
Added support to CloudSlang for dependencies on action level

### LINKED ISSUE #630: Support backward compatibility - versioning
- backward compatibility *may* be affected by the cloudslang versioning plans (e.g. adding version tag in DSL)
- visual data in flows  - [SOLVED](https://github.com/CloudSlang/cloud-slang/issues/643)

### PR DESCRIPTION
**Pull request by genadi-hp**
fixes #630 

Nothing to do with them until we have dependencies resolution. Dependencies for java are still optional for the same reason.

Signed-off-by: Genadi Rabinovich <genadi@hpe.com>

### CODE REVIEW

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformer.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public List<String> transform(List<String> rawData) {
+        return (rawData == null) ? Collections.<String>emptyList() : rawData;
+    }
+
+    @Override
+    public List<Scope> getScopes() {return Arrays.asList(Scope.ACTION);}
+
+    @Override
+    public String keyToTransform() {return SlangTextualKeys.ACTION_DEPENDENCIES;}
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformer.java]: I think it would be better to call the class `ActionDependenciesTransformer` so the scope would be clear, in that case you could even return null here since the key can be taken from class name

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java [hunk 2]
```diff
@@ -55,6 +55,7 @@
     public void doAction(@Param(ScoreLangConstants.RUN_ENV) RunEnvironment runEnv,
                          @Param(ExecutionParametersConsts.NON_SERIALIZABLE_EXECUTION_DATA) Map<String, Object> nonSerializableExecutionData,
                          @Param(ScoreLangConstants.ACTION_TYPE) ActionType actionType,
+                         @Param(ScoreLangConstants.ACTION_DEPENDENCIES) List<String> actionDependencies,
```
Comment (Bonczidai) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java]: best practice: for an ExecutionStep parameter the name (`actionDependencies`) should be the same as the `@Param` value (dependencies), however `@Param` counts

I suggest either `actionDependencies` -> `dependencies` or change `ScoreLangConstants.ACTION_DEPENDENCIES` -> `actionDependencies`

File: cloudslang-compiler/src/test/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformerTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Test
+    public void testTransformWithEmptyDependencies() throws Exception {
+        DependenciesTransformer transformer = new DependenciesTransformer();
+
+        List<String> list = Lists.newArrayList();
+        List<String> transform = transformer.transform(list);
+        assertTrue(transform.isEmpty());
+        assertEquals(list, transform);
+    }
```
Comment (Bonczidai) [inline cloudslang-compiler/src/test/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformerTest.java]: maybe add a test case for transforming non-empty list

**Comment by Bonczidai** [comment]
:+1: very neat @genadi-hp 

added some minor comments, please take a look

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformer.java (hunk 1 above)
Comment (genadi-hp) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformer.java]: This could be a good idea, but there were talks about possibility to have dependencies on upper (operation) level as well. We may want to have multiple scope for dependencies and in this case DependenciesTransformer may be more generic name.

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java (hunk 2 above)
Comment (genadi-hp) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java]: Ok, thanks, will rename actionDependencies -> dependencies

File: cloudslang-compiler/src/test/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformerTest.java (hunk 3 above)
Comment (genadi-hp) [inline cloudslang-compiler/src/test/java/io/cloudslang/lang/compiler/modeller/transformers/DependenciesTransformerTest.java]: Thanks, will be added

**Comment by genadi-hp** [comment]
These changes are no more relevant for us, I am closing this PR
